internal/webhook/external/quay.go

```diff
@@ -0,0 +1,161 @@
+	quay                    = "quay"
+	quayWebhookBodyMaxBytes = 2 << 20 // 2MB
+)
+
+func init() {
+	registry.register(
+		quay,
+		webhookReceiverRegistration{
+			predicate: func(cfg kargoapi.WebhookReceiverConfig) bool {
+				return cfg.Quay != nil
+			},
+			factory: newQuayWebhookReceiver,
+		},
+	)
+}
+
+// quayWebhookReceiver is an implementation of WebhookReceiver that handles
+// inbound webhooks from Quay.
+type quayWebhookReceiver struct {
+	*baseWebhookReceiver
+}
+
+// newQuayWebhookReceiver returns a new instance of quayWebhookReceiver.
+func newQuayWebhookReceiver(
+	c client.Client,
+	project string,
+	cfg kargoapi.WebhookReceiverConfig,
+) WebhookReceiver {
+	return &quayWebhookReceiver{
+		baseWebhookReceiver: &baseWebhookReceiver{
+			client:     c,
+			project:    project,
+			secretName: cfg.Quay.SecretRef.Name,
+		},
+	}
+}
+
+// GetDetails implements WebhookReceiver.
+func (q *quayWebhookReceiver) getReceiverType() string {
+	return quay
+}
+
+// getSecretValues implements WebhookReceiver.
+func (q *quayWebhookReceiver) getSecretValues(
+	_ map[string][]byte,
+) ([]string, error) {
+	// Because Quay does not provide support for a secret
+	// during configuration, the unguessable URL is it's
+	// own implicit authentication mechanism.
+	// So we don't need to do anything here.
+	return nil, nil
+}
+
+// GetHandler implements WebhookReceiver.
+func (q *quayWebhookReceiver) GetHandler() http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		logger := logging.LoggerFromContext(ctx)
+		logger.Debug("identifying source repository")
+
+		if contentLength := r.ContentLength; contentLength > quayWebhookBodyMaxBytes {
+			xhttp.WriteErrorJSON(
+				w,
+				xhttp.Error(
+					fmt.Errorf("content exceeds limit of %d bytes", quayWebhookBodyMaxBytes),
+					http.StatusRequestEntityTooLarge,
+				),
+			)
+			return
+		}
+
+		b, err := io.LimitRead(r.Body, quayWebhookBodyMaxBytes)
+		if err != nil {
+			if errors.Is(err, &io.BodyTooLargeError{}) {
+				xhttp.WriteErrorJSON(
+					w,
+					xhttp.Error(err, http.StatusRequestEntityTooLarge),
+				)
+				return
+			}
+			xhttp.WriteErrorJSON(w, err)
+		}
+
+		payload := struct {
+			// format: quay.io/mynamespace/repository
+			DockerURL string `json:"docker_url"`
+		}{}
+
+		if err = json.Unmarshal(b, &payload); err != nil {
+			xhttp.WriteErrorJSON(
+				w,
+				xhttp.Error(errors.New("invalid request body"),
+					http.StatusBadRequest,
+				),
+			)
+		}
+
+		if payload.DockerURL == "" {
+			xhttp.WriteErrorJSON(w,
+				xhttp.Error(
+					fmt.Errorf("missing repository web URL in request body"),
+					http.StatusBadRequest,
+				),
+			)
+			return
+		}
+
+		logger = logger.WithValues("repoURL", payload.DockerURL)
+		ctx = logging.ContextWithLogger(ctx, logger)
+		result, err := refreshWarehouses(ctx, q.client, q.project, payload.DockerURL)
+		if err != nil {
+			xhttp.WriteErrorJSON(w, err)
+			return
+		}
```

Comment:
Nit: Slightly easier to follow what's going on if we bring a `repoURL` variable into play because it's _the_ thing we put all this effort into finding and Warehouses are indexed by repo URLs. It sort of connects the dots.

```suggestion
		repoURL := payload.DockerURL

		logger = logger.WithValues("repoURL", repoURL)
		ctx = logging.ContextWithLogger(ctx, logger)

		result, err := refreshWarehouses(ctx, q.client, q.project, repoURL)
		if err != nil {
			xhttp.WriteErrorJSON(w, err)
			return
		}
```